README.md
# AnimatedSearchBar [![pub package](https://img.shields.io/pub/v/search_bar_animated.svg)](https://pub.dev/packages/search_bar_animated)

A flutter package that has an animated search bar animation transformation.
It can be fully customized.
![Animated Searchbar Demo](gifs/searchbar.gif)

## Usage Examples

```
 List list = [
   "Flutter",
   "Twitter",
   "Facebook",
   "Google",
   "Instagram",
   "Youtube",
 ];

AnimatedSearchBar(
             shadow: const [
               BoxShadow(
                   color: Colors.black38, blurRadius: 6, offset: Offset(0, 6))
             ],
             backgroundColor: Colors.white,
             buttonColor: Colors.blue,
             width: MediaQuery.of(context).size.width * 0.7,
             submitButtonColor: Colors.blue,
             textStyle: const TextStyle(color: Colors.blue),
             buttonIcon: const Icon(
               Icons.search,
             ),
             duration: const Duration(milliseconds: 1000),
             submitIcon: const Icon(Icons.send),
             animationAlignment: AnimationAlignment.left,
             onSubmit: () {
               setState(() {
                 value = textController.text;
               });
             },
             searchList: list,
             searchQueryBuilder: (query, list) => list.where((item) {
               return item!
                   .toString()
                   .toLowerCase()
                   .contains(query.toLowerCase());
             }).toList(),
             textController: textController,
             overlaySearchListItemBuilder: (dynamic item) => Container(
               padding: const EdgeInsets.all(8),
               child: Text(
                 item,
                 style: const TextStyle(fontSize: 15, color: Colors.black),
               ),
             ),
             onItemSelected: (dynamic item) {
               textController.value = textController.value.copyWith(
                 text: item.toString(),
               );
             },
             overlaySearchListHeight: 100,
           ),
```

## Getting Started

In the `pubspec.yaml` of your flutter project, add the following dependency:

```` 
search_bar_animated: "^latest_version"
````
in your Dart code, you can use:
````
import 'package:search_bar_animated/search_bar_animated.dart';
````

## Available Parameters

| Param                                                                   | isRequired |
| ----------------------------------------------------------------------- | ---------- |
| **List of widgets** searchList                                          | yes        |
| **Function(dynamic)** overlaySearchListItemBuilder                      | yes        |
| **Function(query,list) => list** searchQueryBuilder                     | yes        |
| **Icon** buttonIcon                                                     | yes        |
| **Icon** submitIcon                                                     | yes        |
| **Function** onSubmit                                                   | yes        |
| **Duration** duration, *Default is 1 second*                            | No         |
| **double** width ,      *Default is 300*                                | No         |
| **String** hintText ,    *Default is "Search..."*                       | No         |
| **String** failMessage ,                                                | No         |
| **List of BoxShadow** shadow ,                                          | No         |
| **Controller** textController                                           | No         |
| **AnimationAlignment(left or right)** animationAlignment                | No         |
| **Function(dynamic)** onItemSelected                                    | No         |
| **bool** hideSearchBoxWhenItemSelected    *Default is false*            | No         |
| **double** overlaySearchListHeight                                      | No         |
| **Widget** noItemsFoundWidget                                           | No         |
| **TextStyle** textStyle ,                                               | No         |
| **InputDecoration** searchBoxInputDecoration ,                          | No         |
| **Color** backgroundColor                *Default is white*             | No         |
| **Color** buttonColor ,                  *Default is black*             | No         |
| **Color** buttonBackgroundColor ,        *Default is transparent*       | No         |
| **Color** submitButtonColor ,            *Default is black*             | No         |
| **Color** submitButtonBackgroundColor ,  *Default is transparent*       | No         |

---

<br><br>


CHANGELOG.md
## 0.0.1
* Initial release.

## 1.0.0
* Add miss information

## 1.0.1
* Add example

## 1.0.2
* Fix example


example/lib/home_page.dart

import 'package:flutter/material.dart';
import 'package:search_bar_animated/search_bar_animated.dart';

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  List list = [
    "Flutter",
    "Twitter",
    "Facebook",
    "Google",
    "Instagram",
    "Youtube",
  ];

  TextEditingController textController = TextEditingController();
  String value = "";

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: AnimatedSearchBar(
          backgroundColor: Colors.transparent,
          buttonColor: Colors.white,
          width: MediaQuery.of(context).size.width,
          submitButtonColor: Colors.white,
          textStyle: const TextStyle(color: Colors.white),
          buttonIcon: const Icon(
            Icons.search,
          ),
          duration: const Duration(milliseconds: 700),
          submitIcon: const Icon(Icons.arrow_back_rounded),
          animationAlignment: AnimationAlignment.right,
          onSubmit: () {
            setState(() {
              value = textController.text;
            });
          },
          searchList: list,
          searchQueryBuilder: (query, list) => list.where((item) {
            return item!.toString().toLowerCase().contains(query.toLowerCase());
          }).toList(),
          textController: textController,
          overlaySearchListItemBuilder: (dynamic item) => Container(
            padding: const EdgeInsets.all(8),
            child: Text(
              item,
              style: const TextStyle(fontSize: 15, color: Colors.black),
            ),
          ),
          onItemSelected: (dynamic item) {
            textController.value = textController.value.copyWith(
              text: item.toString(),
            );
          },
          overlaySearchListHeight: 100,
        ),
      ),
      body: Container(
        color: Colors.blueGrey,
        child: Column(
          children: [
            const SizedBox(
              height: 200,
            ),
            AnimatedSearchBar(
              shadow: const [
                BoxShadow(
                    color: Colors.black38, blurRadius: 6, offset: Offset(0, 6))
              ],
              backgroundColor: Colors.white,
              buttonColor: Colors.blue,
              width: MediaQuery.of(context).size.width * 0.7,
              submitButtonColor: Colors.blue,
              textStyle: const TextStyle(color: Colors.blue),
              buttonIcon: const Icon(
                Icons.search,
              ),
              duration: const Duration(milliseconds: 1000),
              submitIcon: const Icon(Icons.send),
              animationAlignment: AnimationAlignment.left,
              onSubmit: () {
                setState(() {
                  value = textController.text;
                });
              },
              searchList: list,
              searchQueryBuilder: (query, list) => list.where((item) {
                return item!
                    .toString()
                    .toLowerCase()
                    .contains(query.toLowerCase());
              }).toList(),
              textController: textController,
              overlaySearchListItemBuilder: (dynamic item) => Container(
                padding: const EdgeInsets.all(8),
                child: Text(
                  item,
                  style: const TextStyle(fontSize: 15, color: Colors.black),
                ),
              ),
              onItemSelected: (dynamic item) {
                textController.value = textController.value.copyWith(
                  text: item.toString(),
                );
              },
              overlaySearchListHeight: 100,
            ),
            const SizedBox(
              height: 50,
            ),
            Center(
              child: Text(
                value,
                style: const TextStyle(fontSize: 30, color: Colors.white),
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/search_bar_animated.dart
import 'dart:math';
import 'package:flutter/material.dart';

typedef QueryListItemBuilder<T> = Widget Function(T item);
typedef OnItemSelected<T> = void Function(T item);
typedef QueryBuilder<T> = List<T> Function(
  String query,
  List<T> list,
);
enum AnimationAlignment { left, right }

class AnimatedSearchBar<T> extends StatefulWidget {
  const AnimatedSearchBar(
      {required this.searchList,
      required this.overlaySearchListItemBuilder,
      required this.searchQueryBuilder,
      required this.buttonIcon,
      required this.submitIcon,
      required this.onSubmit,
      Key? key,
      this.width,
      this.hintText,
      this.duration,
      this.failMessage,
      this.shadow,
      this.textController,
      this.animationAlignment,
      this.onItemSelected,
      this.hideSearchBoxWhenItemSelected = false,
      this.overlaySearchListHeight,
      this.noItemsFoundWidget,
      this.backgroundColor,
      this.buttonBackgroundColor,
      this.submitButtonBackgroundColor,
      this.searchBoxInputDecoration,
      this.buttonColor,
      this.textStyle,
      this.submitButtonColor})
      : super(key: key);

  final List<T> searchList;
  final QueryListItemBuilder<T> overlaySearchListItemBuilder;
  final bool hideSearchBoxWhenItemSelected;
  final double? overlaySearchListHeight;
  final QueryBuilder<T> searchQueryBuilder;
  final Widget? noItemsFoundWidget;
  final OnItemSelected<T>? onItemSelected;
  final InputDecoration? searchBoxInputDecoration;
  final TextEditingController? textController;
  final double? width;
  final Color? backgroundColor;
  final Color? buttonBackgroundColor;
  final Color? submitButtonBackgroundColor;
  final Function onSubmit;
  final List<BoxShadow>? shadow;
  final Widget buttonIcon;
  final Widget submitIcon;
  final String? hintText;
  final Duration? duration;
  final Color? buttonColor;
  final Color? submitButtonColor;
  final TextStyle? textStyle;
  final AnimationAlignment? animationAlignment;
  final String? failMessage;

  @override
  _AnimatedSearchBarState<T> createState() => _AnimatedSearchBarState<T>();
}

class _AnimatedSearchBarState<T> extends State<AnimatedSearchBar<T?>>
    with SingleTickerProviderStateMixin {
  bool toggle = false;
  late List<T> _list;
  late List<T?> _searchList;
  bool? isFocused;
  late FocusNode _focusNode;
  late ValueNotifier<T?> notifier;
  bool? isRequiredCheckFailed;
  Widget? searchBox;
  OverlayEntry? overlaySearchList;
  bool showTextBox = false;
  double? overlaySearchListHeight;
  final LayerLink _layerLink = LayerLink();
  final double textBoxHeight = 30;
  TextEditingController textController = TextEditingController();
  bool isSearchBoxSelected = false;
  late AnimationController _animationController;
  late List<BoxShadow> shadow;
  late String _hintText;
  late Duration duration;

  @override
  void initState() {
    super.initState();
    init();
  }

  void init() {
    _hintText = widget.hintText ?? "Search...";
    duration = widget.duration ?? const Duration(seconds: 1);
    _searchList = <T>[];
    textController = widget.textController ?? textController;
    notifier = ValueNotifier(null);
    _focusNode = FocusNode();
    isFocused = false;
    _list = List<T>.from(widget.searchList);
    _searchList.addAll(_list);
    _focusNode.addListener(() {
      if (!_focusNode.hasFocus) {
        overlaySearchList?.remove();
        overlaySearchList = null;
      } else {
        _searchList
          ..clear()
          ..addAll(_list);
        if (overlaySearchList == null) {
        } else {}
      }
    });
    textController.addListener(() {
      final text = textController.text;
      if (text.trim().isNotEmpty) {
        _searchList.clear();

        final List<T?> filterList =
            widget.searchQueryBuilder(text, widget.searchList);
        _searchList.addAll(filterList);
        if (overlaySearchList == null) {
        } else {
          overlaySearchList?.markNeedsBuild();
        }
      } else {
        _searchList
          ..clear()
          ..addAll(_list);
        if (overlaySearchList == null) {
        } else {
          overlaySearchList?.markNeedsBuild();
        }
      }
    });
    _animationController = AnimationController(vsync: this, duration: duration);
  }

  @override
  void didUpdateWidget(AnimatedSearchBar oldWidget) {
    if (oldWidget.searchList != widget.searchList) {
      init();
    }

    super.didUpdateWidget(oldWidget as AnimatedSearchBar<T>);
  }

  void onSearchListItemSelected(T? item) {
    overlaySearchList?.remove();

    overlaySearchList = null;
    _focusNode.unfocus();
    setState(() {
      notifier.value = item;
      isFocused = false;
      isRequiredCheckFailed = false;
      isSearchBoxSelected = false;
    });
    if (widget.onItemSelected != null) {
      widget.onItemSelected!(item);
    }
  }

  void onTextFieldFocus() {
    setState(() {
      isSearchBoxSelected = true;
    });
    overlaySearchList?.markNeedsBuild();
    final RenderBox searchBoxRenderBox =
        context.findRenderObject() as RenderBox;
    final RenderBox overlay =
        Overlay.of(context)?.context.findRenderObject() as RenderBox;
    final width = searchBoxRenderBox.size.width * 0.8;
    final position = RelativeRect.fromRect(
      Rect.fromPoints(
        searchBoxRenderBox.localToGlobal(
          searchBoxRenderBox.size.topLeft(Offset.zero),
          ancestor: overlay,
        ),
        searchBoxRenderBox.localToGlobal(
          searchBoxRenderBox.size.topRight(Offset.zero),
          ancestor: overlay,
        ),
      ),
      Offset.zero & overlay.size,
    );
    overlaySearchList = OverlayEntry(
        builder: (context) => Positioned(
              left: position.left,
              width: width,
              child: CompositedTransformFollower(
                offset: const Offset(
                  0,
                  40,
                ),
                showWhenUnlinked: false,
                link: _layerLink,
                child: Card(
                  margin: const EdgeInsets.all(12),
                  color: Colors.white,
                  elevation: 5,
                  shape: const RoundedRectangleBorder(
                    borderRadius: BorderRadius.all(Radius.circular(4)),
                  ),
                  child: _searchList.isNotEmpty
                      ? Column(
                          mainAxisAlignment: MainAxisAlignment.end,
                          crossAxisAlignment: CrossAxisAlignment.end,
                          children: <Widget>[
                            SizedBox(
                              height: overlaySearchListHeight,
                              child: Scrollbar(
                                child: ListView.separated(
                                  padding:
                                      const EdgeInsets.symmetric(vertical: 4),
                                  separatorBuilder: (context, index) =>
                                      const Divider(
                                    height: 1,
                                  ),
                                  itemBuilder: (context, index) => Material(
                                    color: Colors.transparent,
                                    child: InkWell(
                                      onTap: () => onSearchListItemSelected(
                                          _searchList[index]),
                                      child:
                                          widget.overlaySearchListItemBuilder(
                                        _searchList.elementAt(index),
                                      ),
                                    ),
                                  ),
                                  itemCount: _searchList.length,
                                ),
                              ),
                            ),
                          ],
                        )
                      : widget.noItemsFoundWidget != null
                          ? Center(
                              child: widget.noItemsFoundWidget,
                            )
                          : Container(
                              padding: const EdgeInsets.only(
                                  top: 7, right: 7, left: 7),
                              height: 30,
                              child: Text(
                                widget.failMessage ?? "",
                                textAlign: TextAlign.start,
                              ),
                            ),
                ),
              ),
            ));
    Overlay.of(context)?.insert(overlaySearchList!);
  }

  @override
  Widget build(BuildContext context) {
    overlaySearchListHeight = widget.overlaySearchListHeight ??
        MediaQuery.of(context).size.height / 4;
    double? width = widget.width != null
        ? (widget.width! < 200)
            ? 200
            : widget.width
        : 300;

    searchBox = Container(
      margin: const EdgeInsets.only(right: 80),
      child: TextField(
        onChanged: (value) {
          if (overlaySearchList == null) {
            onTextFieldFocus();
          } else {
            overlaySearchList?.markNeedsBuild();
          }
        },
        controller: textController,
        focusNode: _focusNode,
        style: widget.textStyle ??
            const TextStyle(fontSize: 16, color: Colors.black),
        decoration: widget.searchBoxInputDecoration ??
            InputDecoration(
              enabledBorder: const OutlineInputBorder(
                borderSide: BorderSide.none,
              ),
              focusedBorder:
                  const OutlineInputBorder(borderSide: BorderSide.none),
              border: InputBorder.none,
              hintText: _hintText,
              hintStyle: widget.textStyle ??
                  const TextStyle(fontSize: 16, color: Colors.black),
              contentPadding: const EdgeInsets.only(
                left: 16,
                right: 20,
                top: 14,
                bottom: 14,
              ),
            ),
      ),
    );

    final searchBoxBody = Column(
      mainAxisSize: MainAxisSize.min,
      crossAxisAlignment: CrossAxisAlignment.start,
      children: <Widget>[
        if (widget.hideSearchBoxWhenItemSelected && notifier.value != null)
          const SizedBox(height: 0)
        else
          CompositedTransformTarget(
            link: _layerLink,
            child: searchBox,
          ),
      ],
    );

    return Container(
      height: 50,
      width: width,
      alignment: widget.animationAlignment != null &&
              widget.animationAlignment == AnimationAlignment.right
          ? const Alignment(1, 0)
          : const Alignment(-1, 0),
      child: AnimatedContainer(
        duration: duration,
        height: 50,
        width: toggle ? width : 50,
        curve: Curves.easeInOut,
        decoration: BoxDecoration(
          color: widget.backgroundColor ?? Colors.blue,
          borderRadius: BorderRadius.circular(30),
          boxShadow: widget.shadow,
        ),
        child: Stack(
          children: [
            AnimatedPositioned(
              top: 1,
              bottom: 1,
              right: widget.animationAlignment != null &&
                      widget.animationAlignment == AnimationAlignment.left
                  ? 1
                  : null,
              left: widget.animationAlignment != null &&
                      widget.animationAlignment == AnimationAlignment.right
                  ? 1
                  : null,
              curve: Curves.easeOut,
              duration: duration,
              child: AnimatedOpacity(
                opacity: toggle ? 1 : 0,
                duration: duration,
                child: AnimatedBuilder(
                  child: Container(
                    decoration: BoxDecoration(
                        color: widget.submitButtonBackgroundColor ??
                            Colors.transparent,
                        borderRadius: BorderRadius.circular(30)),
                    child: Center(
                      child: IconButton(
                        color: widget.submitButtonColor ?? Colors.black,
                        padding: EdgeInsets.zero,
                        iconSize: 25,
                        icon: widget.submitIcon,
                        onPressed: () {
                          widget.onSubmit();
                          setState(() {
                            if (toggle) {
                              _animationController.reverse();
                            } else {
                              _animationController.forward();
                            }
                            toggle = false;
                          });
                        },
                      ),
                    ),
                  ),
                  builder: (context, widget) {
                    return Transform.rotate(
                      angle: _animationController.value * pi * 2,
                      child: widget,
                    );
                  },
                  animation: _animationController,
                ),
              ),
            ),
            AnimatedPositioned(
              left: toggle ? 45 : 10,
              duration: duration,
              child: AnimatedOpacity(
                opacity: toggle ? 1 : 0,
                duration: duration,
                child: SizedBox(
                  width: width! - 60,
                  child: searchBoxBody,
                ),
              ),
            ),
            Positioned(
              top: 1,
              bottom: 1,
              right: widget.animationAlignment != null &&
                      widget.animationAlignment == AnimationAlignment.right
                  ? 1
                  : null,
              left: widget.animationAlignment != null &&
                      widget.animationAlignment == AnimationAlignment.left
                  ? 1
                  : null,
              child: Container(
                decoration: BoxDecoration(
                  color: widget.buttonBackgroundColor ?? Colors.transparent,
                  borderRadius: BorderRadius.circular(30),
                ),
                child: IconButton(
                  iconSize: 30,
                  color: widget.buttonColor ?? Colors.black,
                  onPressed: () {
                    textController.clear();
                    setState(() {
                      if (toggle) {
                        _animationController.reverse();
                      } else {
                        _animationController.forward();
                      }
                      toggle = !toggle;
                    });
                  },
                  icon: widget.buttonIcon,
                ),
              ),
            )
          ],
        ),
      ),
    );
  }
}


example/lib/main.dart

import 'package:flutter/material.dart';
import 'package:search_bar_animated/search_bar_animated.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home:const HomePage(),
    );
  }
}


class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  List list = [
    "Flutter",
    "Twitter",
    "Facebook",
    "Google",
    "Instagram",
    "Youtube",
  ];

  TextEditingController textController = TextEditingController();
  String value = "";

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: AnimatedSearchBar(
          backgroundColor: Colors.transparent,
          buttonColor: Colors.white,
          width: MediaQuery.of(context).size.width,
          submitButtonColor: Colors.white,
          textStyle: const TextStyle(color: Colors.white),
          buttonIcon: const Icon(
            Icons.search,
          ),
          duration: const Duration(milliseconds: 700),
          submitIcon: const Icon(Icons.arrow_back_rounded),
          animationAlignment: AnimationAlignment.right,
          onSubmit: () {
            setState(() {
              value = textController.text;
            });
          },
          searchList: list,
          searchQueryBuilder: (query, list) => list.where((item) {
            return item!.toString().toLowerCase().contains(query.toLowerCase());
          }).toList(),
          textController: textController,
          overlaySearchListItemBuilder: (dynamic item) => Container(
            padding: const EdgeInsets.all(8),
            child: Text(
              item,
              style: const TextStyle(fontSize: 15, color: Colors.black),
            ),
          ),
          onItemSelected: (dynamic item) {
            textController.value = textController.value.copyWith(
              text: item.toString(),
            );
          },
          overlaySearchListHeight: 100,
        ),
      ),
      body: Container(
        color: Colors.blueGrey,
        child: Column(
          children: [
            const SizedBox(
              height: 200,
            ),
            AnimatedSearchBar(
              shadow: const [
                BoxShadow(
                    color: Colors.black38, blurRadius: 6, offset: Offset(0, 6))
              ],
              backgroundColor: Colors.white,
              buttonColor: Colors.blue,
              width: MediaQuery.of(context).size.width * 0.7,
              submitButtonColor: Colors.blue,
              textStyle: const TextStyle(color: Colors.blue),
              buttonIcon: const Icon(
                Icons.search,
              ),
              duration: const Duration(milliseconds: 1000),
              submitIcon: const Icon(Icons.send),
              animationAlignment: AnimationAlignment.left,
              onSubmit: () {
                setState(() {
                  value = textController.text;
                });
              },
              searchList: list,
              searchQueryBuilder: (query, list) => list.where((item) {
                return item!
                    .toString()
                    .toLowerCase()
                    .contains(query.toLowerCase());
              }).toList(),
              textController: textController,
              overlaySearchListItemBuilder: (dynamic item) => Container(
                padding: const EdgeInsets.all(8),
                child: Text(
                  item,
                  style: const TextStyle(fontSize: 15, color: Colors.black),
                ),
              ),
              onItemSelected: (dynamic item) {
                textController.value = textController.value.copyWith(
                  text: item.toString(),
                );
              },
              overlaySearchListHeight: 100,
            ),
            const SizedBox(
              height: 50,
            ),
            Center(
              child: Text(
                value,
                style: const TextStyle(fontSize: 30, color: Colors.white),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
